src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -201,14 +202,47 @@ internal static IEnumerable<string> GetCommaSeparatedValues(string text)
             return validName;
         }
 
-        private static IEnumerable<TagHelperAttributeDescriptor> GetAttributeDescriptors(Type type)
+        private static IEnumerable<TagHelperAttributeDescriptor> GetAttributeDescriptors(
+            Type type, 
+            ErrorSink errorSink)
         {
-            var properties = type.GetRuntimeProperties().Where(IsAccessibleProperty);
-            var attributeDescriptors = properties.Select(ToAttributeDescriptor);
+            var accessibleProperties = type.GetRuntimeProperties().Where(IsAccessibleProperty);
+            var attributeDescriptors = new List<TagHelperAttributeDescriptor>();
+
+            foreach (var property in accessibleProperties)
+            {
+                var descriptor = ToAttributeDescriptor(property);
+                if (ValidateTagHelperAttributeDescriptor(descriptor, type, errorSink))
+                {
+                    attributeDescriptors.Add(descriptor);
+                }
+            }
 
             return attributeDescriptors;
         }
 
+        private static bool ValidateTagHelperAttributeDescriptor(
+            TagHelperAttributeDescriptor attributeDescriptor,
+            Type parentType,
+            ErrorSink errorSink)
+        {
+            // data-* attributes are explicitly not implemented by user agents and are not intended for use on 
+            // the server; therefore it's invalid for TagHelpers to bind to them.
+            if (attributeDescriptor.Name.StartsWith(DataDashPrefix, StringComparison.OrdinalIgnoreCase))
+            {
+                errorSink.OnError(
+                    SourceLocation.Zero, 
```

Comment:
We notify the developer what class/property caused the error via the error message, it's our primary outlet for informing the dev if they screw up when creating their own `TagHelper`s (incredibly unlikely to ever see this error unless you're creating your own `TagHelper`s). 

We *could* have the error light up the `@addTagHelper` on the page to show which directive is the cause of the error but I worry that it may be semi-misleading if the directive is well formed. This is the format we've been following for all the other `TagHelperDescriptor` resolution mechanics but I don't feel strongly one way or another about tying the errors to the `@addTagHelper` directives.